### PR TITLE
An option to write into filesystem cache asynchronously

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
An option to write cache asynchronously (user setting `filesystem_cache_asynchronous_write` and cache settings `background_download_max_memory_usage`).

### CODE REVIEW

File: src/Common/FileCacheSettings.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@ void FileCacheSettings::loadFromConfig(const Poco::Util::AbstractConfiguration &
     enable_cache_hits_threshold = config.getUInt64(config_prefix + ".enable_cache_hits_threshold", REMOTE_FS_OBJECTS_CACHE_ENABLE_HITS_THRESHOLD);
     do_not_evict_index_and_mark_files = config.getUInt64(config_prefix + ".do_not_evict_index_and_mark_files", true);
     allow_to_remove_persistent_segments_from_cache_by_default = config.getUInt64(config_prefix + ".allow_to_remove_persistent_segments_from_cache_by_default", true);
+    background_downlaod_max_memory_usage = config.getUInt64(config_prefix + ".background_downlaod_max_memory_usage", REMOTE_FS_OBJECTS_CACHE_BACKGROUND_DOWNLOAD_MEMORY_USAGE);
```
Comment (nvartolomei) [inline src/Common/FileCacheSettings.cpp:18]: typo: downlaod vs download

File: src/Common/ErrorCodes.cpp [hunk 2]
```diff
@@ -637,6 +637,7 @@
     M(666, CANNOT_USE_CACHE) \
     M(667, NOT_INITIALIZED) \
     M(668, INVALID_STATE) \
+    M(669, BACKGROUND_CACHE_WRITE_WAS_CANCELLED) \
```
Comment (tavplubix) [inline src/Common/ErrorCodes.cpp:640]: What do you think about using existing `UNFINISHED` code? We use it for cancelled background operations like merges and killed mutations. (there's also `QUERY_WAS_CANCELLED`, but it's more specific)

Comment (kssenii) [reply src/Common/ErrorCodes.cpp:640]: > What do you think about using existing UNFINISHED code?

Seems good, will use it. 👍🏻

**Comment by alexey-milovidov** [comment]
It can be too complex to continue.

**Comment by kssenii** [comment]
> It can be too complex to continue.

it is not.

**Comment by alexey-milovidov** [comment]
Is it worth it?

**Pull request closed without merging**